text/0000-light-dom.md

```diff
@@ -0,0 +1,315 @@
+    - if the root node is a shadow root, the stylesheet is inserted as a `<style>` element as the first shadow root child.
+
+> As an optimization, the above algorithm may be substituted by the equivalent usage of [constructable stylesheets](https://developers.google.com/web/updates/2019/02/constructable-stylesheets) in supporting browsers. This means that developers should not rely on the `<style>` elements being inserted at a specific place in the DOM; it's an implementation detail.
+
+It is important to notice that the order in which light DOM components are rendered impacts the order in which stylesheets are injected into the root node and directly influences CSS rule specificity.
+
+#### `:host` and `:host-context()`
+
+Also worth noting is that `:host` and `host-context()` are inserted as-is. Therefore, a light DOM component can style its shadow host. For example:
+
+```css
+/* x-light.css */
+:host {
+  background: blue;
+}
+```
+
+```html
+<x-shadow>
+  #shadow-root
+  <style>
+    :host {
+      background: blue;
+    }
+  </style>
+  <x-light></x-light>
+</x-shadow>
+```
+
+In this case, `<x-light>` is able to style `<x-shadow>` via `:host`. Unlike shadow components, `:host` does not refer to the component but instead to its actual shadow host.
+
+This behavior is according to the specification – `:host` anywhere in the DOM refers to the nearest containing shadow root. At the top (document) level, it does nothing.
+
+#### Synthetic shadow
+
+In the case of synthetic shadow, no attempt is made to scope light DOM styles as described above. The styles are simply inserted as-is, and within a synthetic shadow root they will bleed into the rest of the page, including components using synthetic shadow DOM.
+
+This has some precedent – synthetic shadow always allowed global styles to leak into components (although styles could not leak out of those components). So in a synthetic shadow world, light DOM components' CSS will essentially act like
+`<style>`s inserted into the document `<head>`.
+
+#### Scoped styles
+
+Different approaches to layer style scoping on top have been discussed while designing Light DOM, such as introducing a new file extension for automatic style scoping (`.scoped.css`) or using a `<style scoped>` element in the template. This can be addressed in a future RFC.
+
+### Slots
+
+As mentioned before, the component composition model in LWC is provided by slots. Light DOM will provide the same mental model for developers building Light DOM components.
+
+In Light DOM, `<slot>` will denote the place where the slotted component will be attached. The `<slot>` element itself won't be rendered. The slotted content (or the fallback content) will be flattened to the parent element at runtime.
+
+Since the `<slot>` element itself isn't rendered, adding attributes or event listeners to the `<slot>` element in the template will throw a compiler error.
```

Comment:
This affects builders if it is not a real slot.  This may mean every single builder (LAB is only one of the many builders Salesforce has) has to add special code to process components dragged into another component's light dom slot.

Reply:
Unfortunately it would be impossible to use real slots for light DOM, because slots as defined by the web platform are [a part of shadow DOM](https://developer.mozilla.org/en-US/docs/Web/API/Element/slot).

Whatever we decide on for "light DOM slots," it will not be the same as shadow DOM slots.

Reply:
Thanks. Follow-up question: Today, we have internal LWC container components like tabset and accordion that are exposed in LAB (via an aura wrapper).  ISV LWC container components (which are just components that contain slots that the builder understands) are on the roadmap.  It sounds like the recommendation is for ISVs to build these with Light DOM, since they are general purpose and just meant to house other content.

Reply:
It's really up to each use case. I think we would need to draft a longer document about the strengths and weaknesses of light DOM versus shadow DOM.

For instance, even a tabset component may want to protect its shadow content from being unintentionally styled by the consumer, so shadow DOM may be an appropriate choice there. Also a tabset will inevitably use slots, meaning that there could be performance or timing differences between shadow and light DOM. (E.g. eager versus lazy – a tabset containing 10 slots that only renders 1 might still pay the cost of rendering the other 9, if slots are eager, as they are in native shadow DOM.)

But yes, in general, I think your intuition is correct. Tabsets and accordions don't themselves contain sensitive content, so they could be good candidates for light DOM.

Reply:
I believe the issue is broader than builders: any system that generates LWC components needs to be updated if the LWC syntax differs in how slotted content is provided to a Light DOM-enabled component.

Reply:
Other thoughts:
1. Given the styling/perf differences, we should consider whether some components are configurable by the admin
2. We could use guardrails in the builder to guide admins towards choosing the correct configuration by a) visually showing the difference in preview and b) predicting perf characteristics

Reply:
> any system that generates LWC components needs to be updated

Side note: any system that generates LWC components will need to be updated to respect the `no-shadow` directive anyway.